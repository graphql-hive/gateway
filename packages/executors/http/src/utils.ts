import { createGraphQLError } from '@graphql-tools/utils';

export function createAbortErrorReason() {
  return new Error('Executor was disposed.');
}

export function createGraphQLErrorForAbort(
  reason: any,
  extensions?: Record<string, any>,
) {
  return createGraphQLError('The operation was aborted. reason: ' + reason, {
    extensions,
  });
}

export function createResultForAbort(
  reason: any,
  extensions?: Record<string, any>,
) {
  return {
    errors: [createGraphQLErrorForAbort(reason, extensions)],
  };
}
