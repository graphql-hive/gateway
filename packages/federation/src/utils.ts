import { OBJECT_SUBSCHEMA_SYMBOL } from '@graphql-tools/delegate';
import {
  DirectableObject,
  getDirectiveExtensions,
  MapperKind,
  mapSchema,
  memoize1,
  parseSelectionSet,
} from '@graphql-tools/utils';
import { GraphQLError, GraphQLSchema, Kind, print, SelectionNode, TypeNode } from 'graphql';

export const getArgsFromKeysForFederation = memoize1(
  function getArgsFromKeysForFederation(representations: readonly any[]) {
    return { representations };
  },
);

function getAvailableKeys(
  data: any,
  selections: readonly SelectionNode[],
  typeName: string = data?.__typename,
): {
  keys: (string | symbol)[];
  keyMappings: Record<string, string>;
  keySelections: Record<string, readonly SelectionNode[]>;
} {
  const keys = ['__typename', OBJECT_SUBSCHEMA_SYMBOL];
  const keyMappings: Record<string, string> = {};
  const keySelections: Record<string, readonly SelectionNode[]> = {};
  for (const selection of selections) {
    if (selection.kind === Kind.FIELD) {
      const fieldName = selection.name.value;
      const responseKey = selection.alias?.value || selection.name.value;
      if (fieldName !== responseKey) {
        keyMappings[fieldName] = responseKey;
      }
      keys.push(fieldName);
      if (selection.selectionSet?.selections.length) {
        keySelections[responseKey] = selection.selectionSet.selections;
      }
    } else if (selection.kind === Kind.INLINE_FRAGMENT) {
      if (
        selection.typeCondition &&
        typeName != null &&
        typeName !== selection.typeCondition.name.value
      ) {
        continue;
      }
      if (selection.selectionSet.selections.length > 0) {
        const fragmentKeys = getAvailableKeys(data, selection.selectionSet.selections);
        for (const key of fragmentKeys.keys) {
          if (!keys.includes(key)) {
            keys.push(key);
          }
        }
        for (const [key, sel] of Object.entries(fragmentKeys.keySelections)) {
          if (keySelections[key]) {
            keySelections[key] = [
              ...keySelections[key],
              ...sel,
            ];
          } else {
            keySelections[key] = sel;
          }
        }
      }
    }
  }
  return {
    keys,
    keySelections,
    keyMappings,
  };
}

export function projectDataSelectionSet(
  data: any,
  selections?: readonly SelectionNode[],
  typeName?: string,
): any {
  if (
    data == null ||
    selections == null ||
    !selections?.length
  ) {
    return data;
  }
  if (data instanceof Error) {
    return null;
  }
  if (Array.isArray(data)) {
    return data.map((entry) => projectDataSelectionSet(entry, selections, typeName));
  }
  const {
    keys: availableKeys,
    keySelections,
    keyMappings,
  } = getAvailableKeys(data, selections, typeName);
  const ownKeys = Array.from(new Set([...availableKeys, ...Reflect.ownKeys(data).filter(key => typeof key === 'symbol')]));
  const proxy = new Proxy(data, {
    ownKeys() {
      return ownKeys;
    },
    getOwnPropertyDescriptor(target, prop: string) {
      if (typeof prop === 'symbol') {
        return Reflect.getOwnPropertyDescriptor(target, prop);
      }
      if (ownKeys.includes(prop)) {
        return {
          enumerable: typeof prop === 'string',
          configurable: typeof prop === 'string',
          get() {
            if (prop === '__typename' && typeName) {
              return typeName;
            }
            if (availableKeys.includes(prop)) {
              const actualProp = keyMappings[prop as string] || prop;
              const propData = Reflect.get(target, actualProp);
              if (!(propData instanceof GraphQLError) || propData != null) {
                const selectionSet = keySelections[actualProp];
                if (!selectionSet?.length) {
                  return propData;
                }
                const projected = projectDataSelectionSet(propData, selections);
                return projected;
              }
            } else if (typeof prop === 'symbol') {
              return Reflect.get(target, prop);
            } else if (prop === 'toJSON') {
              return () => proxy;
            }
            return undefined;
          }
        };
      }
    },
    get(target, prop: string, receiver) {
      if (prop === '__typename' && typeName) {
        return typeName;
      }
      if (availableKeys.includes(prop)) {
        const actualProp = keyMappings[prop as string] || prop;
        const propData = Reflect.get(target, actualProp, receiver);
        if (!(propData instanceof GraphQLError) || propData != null) {
          const selectionSet = keySelections[actualProp];
          if (!selectionSet?.length) {
            return propData;
          }
          const projected = projectDataSelectionSet(propData, selections);
          return projected;
        }
      } else if (typeof prop === 'symbol') {
        return Reflect.get(target, prop, receiver);
      } else if (prop === 'toJSON') {
        return () => proxy;
      }
      return undefined;
    },
    set() {
      throw new Error('Cannot set properties on a key object');
    },
  });
  return proxy;
}

export function getKeyFnForFederation(typeName: string, keys: string[]) {
  if (keys.some((key) => key.includes('{') || key.includes('('))) {
    const parsedSelectionSet = parseSelectionSet(`{${keys.join(' ')}}`, {
      noLocation: true,
    });
    return function keyFn(root: any) {
      if (root == null) {
        return root;
      }

      return projectDataSelectionSet(root, parsedSelectionSet.selections, typeName);
    };
  }
  const allKeyProps = keys
    .flatMap((key) => key.trim().split(' '))
    .map((key) => key.trim());
  if (allKeyProps.length === 1) {
    const keyProp = allKeyProps[0]!;
    return function keyFn(root: any) {
      if (root == null) {
        return null;
      }
      const keyPropVal = root[keyProp];
      if (keyPropVal == null) {
        return null;
      }
      return {
        __typename: typeName,
        get [keyProp]() {
          return keyPropVal
        },
      };
    };
  }
  return function keyFn(root: any) {
    if (root == null) {
      return null;
    }
    return new Proxy(root, {
      get(root, prop: string, receiver) {
        if (prop === '__typename') {
          return typeName;
        }
        if (allKeyProps.includes(prop)) {
          const propVal = Reflect.get(root, prop, receiver);
          if (propVal instanceof GraphQLError) {
            return undefined;
          }
          return propVal;
        }
        return undefined;
      },
      set() {
        throw new Error('Cannot set properties on a key object');
      },
    });
  };
}

export function getCacheKeyFnFromKey(key: string) {
  if (key.includes('{') || key.includes('(')) {
    const parsedSelectionSet = parseSelectionSet(`{${key}}`, {
      noLocation: true,
    });
    return function cacheKeyFn(root: any) {
      return JSON.stringify(projectDataSelectionSet(root, parsedSelectionSet.selections));
    };
  }
  const keyTrimmed = key.trim();
  const keys = keyTrimmed.split(' ').map((key) => key.trim());
  if (keys.length > 1) {
    return function cacheKeyFn(root: any) {
      let cacheKeyStr = '';
      for (const key of keys) {
        const keyVal = root[key];
        if (keyVal == null) {
          continue;
        } else if (typeof keyVal === 'object') {
          if (cacheKeyStr) {
            cacheKeyStr += ' ';
          }
          cacheKeyStr += JSON.stringify(keyVal);
        } else {
          if (cacheKeyStr) {
            cacheKeyStr += ' ';
          }
          cacheKeyStr += keyVal;
        }
      }
      return cacheKeyStr;
    };
  }
  return function cacheKeyFn(root: any) {
    const keyVal = root[keyTrimmed];
    if (keyVal == null) {
      return '';
    }
    if (typeof keyVal === 'object') {
      return JSON.stringify(keyVal);
    }
    return keyVal;
  };
}

function hasInaccessible(obj: DirectableObject) {
  return getDirectiveExtensions<{
    inaccessible: {};
  }>(obj)?.inaccessible?.length;
}

export function filterInternalFieldsAndTypes(finalSchema: GraphQLSchema) {
  const internalTypeNameRegexp =
    /^(?:_Entity|_Any|_FieldSet|_Service|link|inaccessible|(?:link__|join__|core__)[\w]*)$/;
  return mapSchema(finalSchema, {
    [MapperKind.DIRECTIVE]: (directive) => {
      if (internalTypeNameRegexp.test(directive.name)) {
        return null;
      }
      return directive;
    },
    [MapperKind.TYPE]: (type) => {
      if (internalTypeNameRegexp.test(type.name) || hasInaccessible(type)) {
        return null;
      }
      return type;
    },
    [MapperKind.FIELD]: (fieldConfig) => {
      if (hasInaccessible(fieldConfig)) {
        return null;
      }
      return fieldConfig;
    },
    [MapperKind.QUERY_ROOT_FIELD]: (fieldConfig, fieldName) => {
      if (fieldName === '_entities' || hasInaccessible(fieldConfig)) {
        return null;
      }
      return fieldConfig;
    },
    [MapperKind.ENUM_VALUE]: (valueConfig) => {
      if (hasInaccessible(valueConfig)) {
        return null;
      }
      return valueConfig;
    },
    [MapperKind.ARGUMENT]: (argConfig) => {
      if (hasInaccessible(argConfig)) {
        return null;
      }
      return argConfig;
    },
  });
}

export function getNamedTypeNode(typeNode: TypeNode) {
  if (typeNode.kind !== Kind.NAMED_TYPE) {
    return getNamedTypeNode(typeNode.type);
  }
  return typeNode;
}
