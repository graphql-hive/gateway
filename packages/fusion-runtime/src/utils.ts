import type {
  Transport,
  TransportContext,
  TransportEntry,
  TransportGetSubgraphExecutor,
  TransportGetSubgraphExecutorOptions,
} from '@graphql-mesh/transport-common';
import type { Logger } from '@graphql-mesh/types';
import {
  isDisposable,
  iterateAsync,
  loggerForExecutionRequest,
  requestIdByRequest,
} from '@graphql-mesh/utils';
import {
  DelegationPlanBuilder,
  MergedTypeResolver,
  Subschema,
} from '@graphql-tools/delegate';
import {
  isAsyncIterable,
  isDocumentNode,
  mapAsyncIterator,
  mapMaybePromise,
  printSchemaWithDirectives,
  type ExecutionRequest,
  type Executor,
  type Maybe,
  type MaybePromise,
} from '@graphql-tools/utils';
import { constantCase } from 'constant-case';
import {
  FragmentDefinitionNode,
  print,
  SelectionNode,
  SelectionSetNode,
  type DocumentNode,
  type ExecutionResult,
  type GraphQLSchema,
} from 'graphql';
import type { GraphQLOutputType, GraphQLResolveInfo } from 'graphql/type';

export type {
  TransportEntry,
  TransportGetSubgraphExecutor,
  TransportGetSubgraphExecutorOptions,
};

export type Transports =
  | {
      [key: string]: MaybePromise<Transport | { default: Transport }>;
    }
  | ((kind: string) => MaybePromise<Transport | { default: Transport }>);

function defaultTransportsGetter(kind: string): Promise<Transport> {
  const moduleName = `@graphql-mesh/transport-${kind}`;
  return mapMaybePromise(import(moduleName), (transport) => {
    if (typeof transport !== 'object') {
      throw new Error(`${moduleName} module does not export an object`);
    }
    if (transport?.default?.getSubgraphExecutor) {
      transport = transport.default;
    }
    if (!transport?.getSubgraphExecutor) {
      throw new Error(
        `${moduleName} module does not export "getSubgraphExecutor"`,
      );
    }
    if (typeof transport?.getSubgraphExecutor !== 'function') {
      throw new Error(
        `${moduleName} module's export "getSubgraphExecutor" is not a function`,
      );
    }
    return transport;
  });
}

function getTransportExecutor({
  transportContext,
  transportEntry,
  subgraphName = '',
  subgraph,
  transports = defaultTransportsGetter,
}: {
  transportContext: TransportContext;
  transportEntry: TransportEntry;
  subgraphName?: string;
  subgraph: GraphQLSchema;
  transports?: Transports;
}): MaybePromise<Executor> {
  // TODO
  const kind = transportEntry?.kind || '';
  let logger = transportContext?.logger;
  if (logger) {
    if (subgraphName) {
      logger = logger.child(subgraphName);
    }
    logger?.debug(`Loading transport "${kind}"`);
  }
  return mapMaybePromise(
    typeof transports === 'function' ? transports(kind) : transports[kind],
    (transport) => {
      if (!transport) {
        throw new Error(`Transport "${kind}" is empty`);
      }
      if (typeof transport !== 'object') {
        throw new Error(`Transport "${kind}" is not an object`);
      }
      let getSubgraphExecutor: TransportGetSubgraphExecutor | undefined;
      if ('default' in transport) {
        getSubgraphExecutor = transport.default?.getSubgraphExecutor;
      } else {
        getSubgraphExecutor = transport.getSubgraphExecutor;
      }
      if (!getSubgraphExecutor) {
        throw new Error(
          `Transport "${kind}" does not have "getSubgraphExecutor"`,
        );
      }
      if (typeof getSubgraphExecutor !== 'function') {
        throw new Error(
          `Transport "${kind}" "getSubgraphExecutor" is not a function`,
        );
      }
      return getSubgraphExecutor({
        subgraphName,
        subgraph,
        transportEntry,
        getTransportExecutor(transportEntry) {
          return getTransportExecutor({
            transportContext,
            transportEntry,
            subgraphName,
            subgraph,
            transports,
          });
        },
        ...transportContext,
      });
    },
  );
}

export const subgraphNameByExecutionRequest = new WeakMap<
  ExecutionRequest,
  string
>();

/**
 * This function creates a executor factory that uses the transport packages,
 * and wraps them with the hooks
 */
export function getOnSubgraphExecute({
  onSubgraphExecuteHooks,
  transportContext = {},
  transportEntryMap,
  getSubgraphSchema,
  transportExecutorStack,
  transports,
}: {
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  transports?: Transports;
  transportContext?: TransportContext;
  transportEntryMap: Record<string, TransportEntry>;
  getSubgraphSchema(subgraphName: string): GraphQLSchema;
  transportExecutorStack: AsyncDisposableStack;
}) {
  const subgraphExecutorMap = new Map<string, Executor>();
  return function onSubgraphExecute(
    subgraphName: string,
    executionRequest: ExecutionRequest,
  ) {
    subgraphNameByExecutionRequest.set(executionRequest, subgraphName);
    let executor = subgraphExecutorMap.get(subgraphName);
    // If the executor is not initialized yet, initialize it
    if (executor == null) {
      let logger = transportContext?.logger;
      if (logger) {
        const requestId = requestIdByRequest.get(
          executionRequest.context?.request,
        );
        if (requestId) {
          logger = logger.child(requestId);
        }
        if (subgraphName) {
          logger = logger.child(subgraphName);
        }
        logger.debug(`Initializing executor`);
      }
      // Lazy executor that loads transport executor on demand
      executor = function lazyExecutor(subgraphExecReq: ExecutionRequest) {
        return mapMaybePromise(
          // Gets the transport executor for the given subgraph
          getTransportExecutor({
            transportContext,
            subgraphName,
            get subgraph() {
              return getSubgraphSchema(subgraphName);
            },
            get transportEntry() {
              return transportEntryMap[subgraphName]!;
            },
            transports,
          }),
          (executor_) => {
            if (isDisposable(executor_)) {
              transportExecutorStack.use(executor_);
            }
            // Wraps the transport executor with hooks
            executor = wrapExecutorWithHooks({
              executor: executor_,
              onSubgraphExecuteHooks,
              subgraphName,
              transportEntryMap,
              transportContext,
              getSubgraphSchema,
            });
            // Caches the executor for future use
            subgraphExecutorMap.set(subgraphName, executor);
            return executor(subgraphExecReq);
          },
        );
      };
      // Caches the lazy executor to prevent race conditions
      subgraphExecutorMap.set(subgraphName, executor);
    }
    return executor(executionRequest);
  };
}

export interface WrapExecuteWithHooksOptions {
  executor: Executor;
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  subgraphName: string;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema: (subgraphName: string) => GraphQLSchema;
  transportContext?: TransportContext;
}

declare module 'graphql' {
  interface GraphQLResolveInfo {
    executionRequest?: ExecutionRequest;
  }
}

/**
 * This function wraps the executor created by the transport package
 * with `onSubgraphExecuteHooks` to hook into the execution phase of subgraphs
 */
export function wrapExecutorWithHooks({
  executor,
  onSubgraphExecuteHooks,
  subgraphName,
  transportEntryMap,
  getSubgraphSchema,
  transportContext,
}: WrapExecuteWithHooksOptions): Executor {
  return function executorWithHooks(executionRequest: ExecutionRequest) {
    executionRequest.info = executionRequest.info || ({} as GraphQLResolveInfo);
    executionRequest.info.executionRequest = executionRequest;
    const requestId =
      executionRequest.context?.request &&
      requestIdByRequest.get(executionRequest.context.request);
    let execReqLogger = transportContext?.logger;
    if (execReqLogger) {
      if (requestId) {
        execReqLogger = execReqLogger.child(requestId);
      }
      loggerForExecutionRequest.set(executionRequest, execReqLogger);
    }
    execReqLogger = execReqLogger?.child?.(subgraphName);
    if (onSubgraphExecuteHooks.length === 0) {
      return executor(executionRequest);
    }
    const onSubgraphExecuteDoneHooks: OnSubgraphExecuteDoneHook[] = [];
    return mapMaybePromise(
      iterateAsync(
        onSubgraphExecuteHooks,
        (onSubgraphExecuteHook) =>
          onSubgraphExecuteHook({
            get subgraph() {
              return getSubgraphSchema(subgraphName);
            },
            subgraphName,
            get transportEntry() {
              return transportEntryMap?.[subgraphName];
            },
            executionRequest,
            setExecutionRequest(newExecutionRequest) {
              execReqLogger?.debug(
                'Updating execution request to: ',
                newExecutionRequest,
              );
              executionRequest = newExecutionRequest;
            },
            executor,
            setExecutor(newExecutor) {
              execReqLogger?.debug('executor has been updated');
              executor = newExecutor;
            },
            requestId,
            logger: execReqLogger,
          }),
        onSubgraphExecuteDoneHooks,
      ),
      () => {
        if (onSubgraphExecuteDoneHooks.length === 0) {
          return executor(executionRequest);
        }
        return mapMaybePromise(executor(executionRequest), (currentResult) => {
          const executeDoneResults: OnSubgraphExecuteDoneResult[] = [];
          return mapMaybePromise(
            iterateAsync(
              onSubgraphExecuteDoneHooks,
              (onSubgraphExecuteDoneHook) =>
                onSubgraphExecuteDoneHook({
                  result: currentResult,
                  setResult(newResult: ExecutionResult) {
                    execReqLogger?.debug('overriding result with: ', newResult);
                    currentResult = newResult;
                  },
                }),
              executeDoneResults,
            ),
            () => {
              if (!isAsyncIterable(currentResult)) {
                return currentResult;
              }

              if (executeDoneResults.length === 0) {
                return currentResult;
              }

              const onNextHooks: OnSubgraphExecuteDoneResultOnNext[] = [];
              const onEndHooks: OnSubgraphExecuteDoneResultOnEnd[] = [];

              for (const executeDoneResult of executeDoneResults) {
                if (executeDoneResult.onNext) {
                  onNextHooks.push(executeDoneResult.onNext);
                }
                if (executeDoneResult.onEnd) {
                  onEndHooks.push(executeDoneResult.onEnd);
                }
              }

              if (onNextHooks.length === 0 && onEndHooks.length === 0) {
                return currentResult;
              }

              return mapAsyncIterator(
                currentResult,
                (currentResult) =>
                  mapMaybePromise(
                    iterateAsync(onNextHooks, (onNext) =>
                      onNext({
                        result: currentResult,
                        setResult: (res) => {
                          execReqLogger?.debug('overriding result with: ', res);

                          currentResult = res;
                        },
                      }),
                    ),
                    () => currentResult,
                  ),
                undefined,
                () =>
                  onEndHooks.length === 0
                    ? undefined
                    : iterateAsync(onEndHooks, (onEnd) => onEnd()),
              );
            },
          );
        });
      },
    );
  };
}

export interface UnifiedGraphPlugin<TContext> {
  onSubgraphExecute?: OnSubgraphExecuteHook<TContext>;
  onDelegationPlan?: OnDelegationPlanHook<TContext>;
  onDelegationStageExecute?: OnDelegationStageExecuteHook<TContext>;
}

export type OnSubgraphExecuteHook<TContext = any> = (
  payload: OnSubgraphExecutePayload<TContext>,
) => MaybePromise<Maybe<OnSubgraphExecuteDoneHook | void>>;

export interface OnSubgraphExecutePayload<TContext> {
  subgraph: GraphQLSchema;
  subgraphName: string;
  transportEntry?: TransportEntry;
  executionRequest: ExecutionRequest<any, TContext>;
  setExecutionRequest(executionRequest: ExecutionRequest): void;
  executor: Executor;
  setExecutor(executor: Executor): void;
  requestId?: string;
  logger?: Logger;
}

export interface OnSubgraphExecuteDonePayload {
  result: AsyncIterable<ExecutionResult> | ExecutionResult;
  setResult(result: AsyncIterable<ExecutionResult> | ExecutionResult): void;
}

export type OnSubgraphExecuteDoneHook = (
  payload: OnSubgraphExecuteDonePayload,
) => MaybePromise<Maybe<OnSubgraphExecuteDoneResult | void>>;

export type OnSubgraphExecuteDoneResultOnNext = (
  payload: OnSubgraphExecuteDoneOnNextPayload,
) => MaybePromise<void>;

export interface OnSubgraphExecuteDoneOnNextPayload {
  result: ExecutionResult;
  setResult(result: ExecutionResult): void;
}

export type OnSubgraphExecuteDoneResultOnEnd = () => MaybePromise<void>;

export type OnSubgraphExecuteDoneResult = {
  onNext?: OnSubgraphExecuteDoneResultOnNext;
  onEnd?: OnSubgraphExecuteDoneResultOnEnd;
};

export type OnDelegationPlanHook<TContext> = (
  payload: OnDelegationPlanHookPayload<TContext>,
) => Maybe<OnDelegationPlanDoneHook | void>;

export interface OnDelegationPlanHookPayload<TContext> {
  supergraph: GraphQLSchema;
  subgraph: string;
  sourceSubschema: Subschema<any, any, any, TContext>;
  typeName: string;
  variables: Record<string, any>;
  fragments: Record<string, FragmentDefinitionNode>;
  fieldNodes: SelectionNode[];
  context: TContext;
  requestId?: string;
  logger?: Logger;
  info?: GraphQLResolveInfo;
  delegationPlanBuilder: DelegationPlanBuilder;
  setDelegationPlanBuilder(delegationPlanBuilder: DelegationPlanBuilder): void;
}

export type OnDelegationPlanDoneHook = (
  payload: OnDelegationPlanDonePayload,
) => Maybe<void>;

export interface OnDelegationPlanDonePayload {
  delegationPlan: ReturnType<DelegationPlanBuilder>;
  setDelegationPlan: (
    delegationPlan: ReturnType<DelegationPlanBuilder>,
  ) => void;
}

export type OnDelegationStageExecuteHook<TContext> = (
  payload: OnDelegationStageExecutePayload<TContext>,
) => Maybe<OnDelegationStageExecuteDoneHook>;

export interface OnDelegationStageExecutePayload<TContext> {
  object: any;
  context: TContext;
  info: GraphQLResolveInfo;
  subgraph: string;
  subschema: Subschema<any, any, any, TContext>;
  selectionSet: SelectionSetNode;
  key?: any;
  type: GraphQLOutputType;

  resolver: MergedTypeResolver<TContext>;
  setResolver: (resolver: MergedTypeResolver<TContext>) => void;

  typeName: string;

  requestId?: string;
  logger?: Logger;
}

export type OnDelegationStageExecuteDoneHook = (
  payload: OnDelegationStageExecuteDonePayload,
) => void;

export interface OnDelegationStageExecuteDonePayload {
  result: any;
  setResult: (result: any) => void;
}

export function compareSchemas(
  a: DocumentNode | string | GraphQLSchema,
  b: DocumentNode | string | GraphQLSchema,
) {
  let aStr: string;
  if (typeof a === 'string') {
    aStr = a;
  } else if (isDocumentNode(a)) {
    aStr = print(a);
  } else {
    aStr = printSchemaWithDirectives(a);
  }
  let bStr: string;
  if (typeof b === 'string') {
    bStr = b;
  } else if (isDocumentNode(b)) {
    bStr = print(b);
  } else {
    bStr = printSchemaWithDirectives(b);
  }
  return aStr === bStr;
}

// TODO: Fix this in GraphQL Tools
export function compareSubgraphNames(name1: string, name2: string) {
  return constantCase(name1) === constantCase(name2);
}

export function wrapMergedTypeResolver<TContext extends Record<string, any>>(
  originalResolver: MergedTypeResolver<TContext>,
  typeName: string,
  onDelegationStageExecuteHooks: OnDelegationStageExecuteHook<TContext>[],
  baseLogger?: Logger,
): MergedTypeResolver<TContext> {
  return (object, context, info, subschema, selectionSet, key, type) => {
    let logger = baseLogger;
    let requestId: string | undefined;
    if (logger && context['request']) {
      requestId = requestIdByRequest.get(context['request']);
      if (requestId) {
        logger = logger.child(requestId);
      }
    }
    if (subschema.name) {
      logger = logger?.child(subschema.name);
    }
    let resolver = originalResolver as MergedTypeResolver<TContext>;
    function setResolver(newResolver: MergedTypeResolver<TContext>) {
      resolver = newResolver;
    }
    const onDelegationStageExecuteDoneHooks: OnDelegationStageExecuteDoneHook[] =
      [];
    for (const onDelegationStageExecute of onDelegationStageExecuteHooks) {
      const onDelegationStageExecuteDone = onDelegationStageExecute({
        object,
        context: context as TContext,
        info,
        subgraph: subschema.name!,
        subschema: subschema as Subschema<any, any, any, TContext>,
        selectionSet,
        key,
        typeName,
        type,
        requestId,
        logger,
        resolver,
        setResolver,
      });
      if (onDelegationStageExecuteDone) {
        onDelegationStageExecuteDoneHooks.push(onDelegationStageExecuteDone);
      }
    }
    return mapMaybePromise(
      resolver(
        object,
        context as TContext,
        info,
        subschema as Subschema<any, any, any, TContext>,
        selectionSet,
        key,
        type,
      ),
      (result) => {
        function setResult(newResult: any) {
          result = newResult;
        }
        for (const onDelegationStageExecuteDone of onDelegationStageExecuteDoneHooks) {
          onDelegationStageExecuteDone({
            result,
            setResult,
          });
        }
        return result;
      },
    );
  };
}
