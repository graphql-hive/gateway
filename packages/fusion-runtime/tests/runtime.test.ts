import {
  OnDelegationPlanDoneHook,
  OnDelegationPlanHook,
} from '@graphql-mesh/fusion-runtime';
import { Subschema } from '@graphql-tools/delegate';
import { buildSubgraphSchema } from '@graphql-tools/federation';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { GraphQLSchema, Kind, parse, print } from 'graphql';
import { describe, expect, it, vitest } from 'vitest';
import { composeAndGetExecutor } from './utils';

describe('handleFederationSubschema', () => {
  it('combine federation merging and custom merging', async () => {
    const users = buildSubgraphSchema({
      typeDefs: parse(/* GraphQL */ `
        directive @merge(keyField: String) on FIELD_DEFINITION

        type Query {
          userById(id: ID!): User @merge(keyField: "id")
        }

        type User {
          id: ID!
          name: String
          posts: [Post]
        }

        type Post @key(fields: "id") {
          id: ID!
          author: User
        }
      `),
      resolvers: {
        Query: {
          userById(_root, { id }) {
            return { id, name: `User ${id}` };
          },
        },
        Post: {
          __resolveReference(post) {
            return { id: post.id };
          },
          author(post) {
            return { id: post.id, name: `User ${post.id}` };
          },
        },
        User: {
          posts() {
            return [{ id: '1' }];
          },
        },
      },
    });
    const posts = buildSubgraphSchema({
      typeDefs: parse(/* GraphQL */ `
        type Query {
          posts: [Post]
        }

        type Post @key(fields: "id") {
          id: ID!
          title: String
        }
      `),
      resolvers: {
        Query: {
          posts() {
            return [{ id: '1', title: 'Post 1' }];
          },
        },
        Post: {
          __resolveReference(post) {
            return { id: post.id };
          },
          title(post) {
            return `Post ${post.id}`;
          },
        },
      },
    });
    const executor = composeAndGetExecutor([
      {
        schema: users,
        name: 'users',
      },
      {
        schema: posts,
        name: 'posts',
      },
    ]);
    const result = await executor({
      query: /* GraphQL */ `
        query {
          userById(id: "1") {
            id
            name
            posts {
              id
              title
              author {
                id
                name
              }
            }
          }
          posts {
            id
            title
            author {
              id
              name
            }
          }
        }
      `,
    });
    expect(result).toMatchObject({
      userById: {
        id: '1',
        name: 'User 1',
        posts: [
          {
            id: '1',
            title: 'Post 1',
            author: {
              id: '1',
              name: 'User 1',
            },
          },
        ],
      },
      posts: [
        {
          id: '1',
          title: 'Post 1',
          author: {
            id: '1',
            name: 'User 1',
          },
        },
      ],
    });
  });
});

describe('onDelegationPlanHook', () => {
  it('should be called with the plan', async () => {
    const user = buildSubgraphSchema({
      typeDefs: parse(/* GraphQL */ `
        type Query {
          users: [User]
        }

        type User @key(fields: "id") {
          id: ID!
          name: String
        }
      `),
      resolvers: {
        Query: {
          users() {
            return [{ id: '1', name: 'User 1' }];
          },
        },
        User: {
          __resolveReference(user) {
            return { id: user.id };
          },
          name(user) {
            return `User ${user.id}`;
          },
        },
      },
    });
    const posts = buildSubgraphSchema({
      typeDefs: parse(/* GraphQL */ `
        type Query {
          posts: [Post]
        }

        type Post @key(fields: "id") {
          id: ID!
          title: String
          author: User
        }

        extend type User @key(fields: "id") {
          id: ID! @external
        }
      `),
      resolvers: {
        Query: {
          posts() {
            return [
              { id: '1', authorId: 1 },
              { id: '2', authorId: 1 },
            ];
          },
        },
        Post: {
          __resolveReference(post) {
            return { id: post.id, authorId: 1 };
          },
          title(post) {
            return `Post ${post.id}`;
          },
          author(post) {
            return { id: post.authorId };
          },
        },
      },
    });
    const onDelegationPlanDoneHook = vitest.fn<OnDelegationPlanDoneHook>();
    const context = {
      myContextValue: 'myContextValue',
    };
    const onDelegationPlanHook = vitest.fn<
      OnDelegationPlanHook<typeof context>
    >(() => onDelegationPlanDoneHook);
    const executor = composeAndGetExecutor(
      [
        {
          schema: user,
          name: 'user',
        },
        {
          schema: posts,
          name: 'posts',
        },
      ],
      {
        onDelegationPlanHooks: [onDelegationPlanHook],
      },
    );
    const result = await executor({
      query: /* GraphQL */ `
        query {
          posts {
            id
            title
            author {
              id
              name
            }
          }
        }
      `,
      context,
    });
    expect(onDelegationPlanHook).toHaveBeenCalledTimes(1);
    const onDelegationPlanPayload = onDelegationPlanHook.mock.calls[0]![0];
    expect(onDelegationPlanPayload).toEqual({
      supergraph: expect.any(GraphQLSchema),
      subgraph: expect.any(String),
      sourceSubschema: expect.any(Subschema),
      typeName: expect.any(String),
      variables: expect.any(Object),
      fragments: {},
      fieldNodes: expect.arrayContaining([
        expect.objectContaining({
          kind: Kind.FIELD,
        }),
      ]),
      context,
      delegationPlanBuilder: expect.any(Function),
      setDelegationPlanBuilder: expect.any(Function),
      logger: undefined,
      info: expect.any(Object),
    });
    expect(
      printSchemaWithDirectives(onDelegationPlanPayload.supergraph),
    ).toMatchSnapshot('onDelegationPlanPayload.supergraph');
    expect(onDelegationPlanPayload.subgraph).toMatchSnapshot(
      'onDelegationPlanPayload.subgraph',
    );
    expect(onDelegationPlanPayload.typeName).toMatchSnapshot(
      'onDelegationPlanPayload.typeName',
    );
    expect(onDelegationPlanPayload.variables).toMatchSnapshot(
      'onDelegationPlanPayload.variables',
    );
    expect(print(onDelegationPlanPayload.fieldNodes[0]!)).toMatchSnapshot(
      'onDelegationPlanPayload.fieldNodes[0]',
    );
    expect(result).toEqual({
      posts: [
        {
          id: '1',
          title: 'Post 1',
          author: {
            id: '1',
            name: 'User 1',
          },
        },
        {
          id: '2',
          title: 'Post 2',
          author: {
            id: '1',
            name: 'User 1',
          },
        },
      ],
    });
  });
});
