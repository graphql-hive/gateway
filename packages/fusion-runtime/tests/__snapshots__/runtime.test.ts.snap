// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onDelegationPlanHook should be called with the plan: onDelegationPlanPayload.fieldNodes[0] 1`] = `
"author {
  id
  name
}"
`;

exports[`onDelegationPlanHook should be called with the plan: onDelegationPlanPayload.subgraph 1`] = `"posts"`;

exports[`onDelegationPlanHook should be called with the plan: onDelegationPlanPayload.supergraph 1`] = `
"schema {
  query: Query
}

type Query {
  posts: [Post]
  users: [User]
}

type Post {
  id: ID!
  title: String
  author: User
}

type User {
  id: ID!
  name: String
}"
`;

exports[`onDelegationPlanHook should be called with the plan: onDelegationPlanPayload.typeName 1`] = `"User"`;

exports[`onDelegationPlanHook should be called with the plan: onDelegationPlanPayload.variables 1`] = `{}`;
